### PR TITLE
sqlccl: add a new KeyRewriter abstraction

### PR DESCRIPTION
**Pull request by danhhz**
KeyRewriter is an matcher for an ordered list of pairs of byte prefix
rewrite rules. This will make it trivial to port key rewriting to c++
and will scale more naturally to rewriting interleaved data.

    BenchmarkKeyRewriter-8       100000000            19.4 ns/op           0 B/op           0 allocs/op

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13089)

### CODE REVIEW

File: pkg/ccl/sqlccl/key_rewriter.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	newDesc := *desc
+	newDesc.ID = newTableID
+
+	prefixes := make(map[string]struct{})
+
+	var kr storageccl.KeyRewriter
+	for _, index := range desc.AllNonDropIndexes() {
+		oldPrefix := roachpb.Key(sqlbase.MakeIndexKeyPrefix(desc, index.ID))
+		newPrefix := roachpb.Key(sqlbase.MakeIndexKeyPrefix(&newDesc, index.ID))
+		if _, ok := prefixes[string(oldPrefix)]; !ok {
```
Comment (mjibson) [inline pkg/ccl/sqlccl/key_rewriter.go]: What can cause a duplicate oldPrefix to be encountered?

**Comment by danhhz** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/ccl/sqlccl/key_rewriter.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089#-KbB687McFwDDCElprFn-r1-52:-KbCAV6rzYEa11vOiv3N:b-l4396t) ([raw file](https://github.com/cockroachdb/cockroach/blob/34c86f80f7906b6d8f5a35190baa5af14229d3c7/pkg/ccl/sqlccl/key_rewriter.go#L52)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What can cause a duplicate oldPrefix to be encountered?
</blockquote></details>

Added a comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/key_rewriter.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089#-KbFpevigXQuyGfjJMDh:-KbFpevigXQuyGfjJMDi:b-2or348) ([raw file](https://github.com/cockroachdb/cockroach/blob/3869eede8b07838459218d0032e8baa335d55536/pkg/ccl/storageccl/key_rewriter.go#L32)):*
> ```Go
> 			}
> 			// TODO(dan): Special case when key's cap() is enough.
> 			key = append(rewrite.NewPrefix, key[len(rewrite.OldPrefix):]...)
> ```

If `NewPrefix` happens to have enough extra capacity, we may be reusing that space and overwriting any previous key rewritten using the same rule. We should just allocate the whole thing and copy the two pieces, or append to `rewrite.NewPrefix[:len(rewrite.NewPrefix):len(rewrite.NewPrefix)]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089)*

**Comment by danhhz** [comment]
RFAL

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/key_rewriter.go, line 32 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089#-KbFpevigXQuyGfjJMDh:-KbG9XWXAdHPqd7febzS:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3869eede8b07838459218d0032e8baa335d55536/pkg/ccl/storageccl/key_rewriter.go#L32)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

If `NewPrefix` happens to have enough extra capacity, we may be reusing that space and overwriting any previous key rewritten using the same rule. We should just allocate the whole thing and copy the two pieces, or append to `rewrite.NewPrefix[:len(rewrite.NewPrefix):len(rewrite.NewPrefix)]`
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089#-:-KbG9bwHium4LkeUGkQ9:bb72pd0)*

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13089#-:-KbGBKcMBNREb6pAH9dH:bnfp4nl)*

**Review by dt — APPROVED** [review]

**Comment by danhhz** [comment]
Thanks for the reviews!

**Pull request merged**